[libc++] Fix missing 'get_new_handler()' for Windows builds

siradam7th opened the pull request:
This PR adds the missing `get_new_handler()` for Windows since `<new.h>` in the UCRT is missing this function and only defines `set_new_handler()` so it is best if all of them are defined and implemented by libc++.

Fixes #64813, and also unblocks building for libc++'s C++23 standard library modules on Windows using clang-cl.

siradam7th committed:
[libc++] fix missing 'get_new_handler()' for Windows builds

siradam7th committed:
[libc++] remove 'FIXME' and 'XFAIL' for 'get_new_handler()' test

frederick-vs-ja commented on libcxx/src/new_handler.cpp:
```diff
@@ -11,9 +11,7 @@
-#  if !defined(_LIBCPP_ABI_VCRUNTIME)
-#    define _LIBPCPP_DEFINE_NEW_HANDLER
-#  endif
+#  define _LIBPCPP_DEFINE_NEW_HANDLER
```
I wonder whether we should include `<vcruntime_new.h>` instead of `<new.h>`.

siradam7th replied:
> I wonder whether we should include `<vcruntime_new.h>` instead of `<new.h>` (in other headers).

Unfortunately that would increase the scope of this PR, which is why it exists in standalone form rather than with https://github.com/llvm/llvm-project/pull/14899

siradam7th replied:
> This doesn't seem right. The definitions provided `__new_handler` are dummy and non-conformant. Perhaps we should rely on VCRuntime's functionalities.

Care to elaborate? from what I've seen in MS-STL they both have similar implementations.

mccakit replied:
@frederick-vs-ja Can't we bundle a fixme comment, or bundle a documentation commit to this PR?
When UCRT fixes this issue, this commit can be reverted.

Windows support for module usage is very important for module adoption, waiting for ucrt for this could really slow things down

siradam7th replied:
> Is it really a good idea to include UCRT's `<new.h>`? MSVC STL has been avoiding `<new.h>` for a long while, at least since open-sourcing (see [this](https://github.com/microsoft/STL/blob/219514876ea86491de191ceaa88632616bbc0f19/stl/inc/new)). I think a cleaner way is to follow MSVC STL's strategy - including `<vcruntime_new.h>` and manually declaring `new_handler`, `set_new_handler`, and `get_new_handler`.

I've implemented something similar, it's not ready yet, I have to debug a problem with the handler not being called, once I figure it out I'll commit it here.

siradam7th committed:
[libc++] implement 'set_new_handler' and 'get_new_handler' for Windows

siradam7th committed:
[libc++] replace 'typedef' with a type alias

philnik777 commented on libcxx/include/__new/global_new_delete.h:
```diff
@@ -32,7 +32,7 @@
 
 #if defined(_LIBCPP_ABI_VCRUNTIME)
-#  include <new.h>
+#  include <vcruntime_new.h>
```
Why did this include change?

siradam7th commented on libcxx/include/__new/global_new_delete.h:
```diff
@@ -62,8 +62,8 @@ _LIBCPP_OVERRIDABLE_FUNC_VIS void operator delete(void* __p, std::align_val_t, c
-[[__nodiscard__]] _LIBCPP_OVERRIDABLE_FUNC_VIS void*
-operator new[](std::size_t __sz, std::align_val_t) _THROW_BAD_ALLOC;
+[[__nodiscard__]]
+_LIBCPP_OVERRIDABLE_FUNC_VIS void* operator new[](std::size_t __sz, std::align_val_t) _THROW_BAD_ALLOC;
```
I accidently ran the wrong formatter earlier that did nothing, and I didn't want to interrupt the CI since it was taking a long time already.

philnik777 replied:
This hasn't been addressed.

winkler283 commented:
> The following PR fixes the missing `get_new_handler()` function for Windows by using the same implementation as other platforms since `<new.h>` in the UCRT is missing this function and only defines `set_new_handler()`

 `<new.h>` in the UCRT not only contains `_set_new_handler()` it also contains `_query_new_handler()` which returns a pointer to the current new handler. Isn't it possible to use that function instead of maintaining your own implementation?

siradam7th commented on libcxx/src/new_handler.cpp:
```diff
@@ -26,11 +24,33 @@
+#  else
+  return __libcpp_atomic_exchange(&__new_handler, handler);
+#  endif
+}
```
> Why are we implementing this now? AFAICT the point of the previous implementation was that we shared the new handler with MSVC, which this doesn't do anymore IIUC.

their implementation does not define a `get_new_handler()`.

frederick-vs-ja replied:
> AFAICT the point of the previous implementation was that we shared the new handler with MSVC, which this doesn't do anymore IIUC.

I think that was the point too. But I gradually believe that it made a wrong assumption.

frederick-vs-ja reviewed (approved):
LGTM. But other reviewers possibly have different opinions.

Zingam commented:
@siradam7th Thanks for udating the test. Please be minful of https://github.com/llvm/llvm-project/pull/173451 to merge properly if it lands first.

Also I see an error message about your email address. The project requires to make your  email public before landing. Please make sure you have addressed that.

philnik777 reviewed (changes requested):
If we do this it definitely requires a release note. That should include what implications this change has w.r.t. STL compatibility.

siradam7th commented:
> If we do this it definitely requires a release note. That should include what implications this change has w.r.t. STL compatibility.

@philnik777 care to elaborate more on this? 
which part of these changes might have implications in regards to compatibility?

petrhosek reviewed (approved):
I attempted to implement this in https://reviews.llvm.org/D158450 (that change also has some additional context) and I'm glad someone picked this up.

siradam7th commented:
> > > If we do this it definitely requires a release note. That should include what implications this change has w.r.t. STL compatibility.
> > 
> > 
> > @philnik777 care to elaborate more on this? which part of these changes might have implications in regards to compatibility?
> 
> How did this interact with the MSVC STL until now? How does the behaviour change? Why did we not make it compatible with MSVC? All the questions a potential ABI break implies.
> 
> Looking at this all again, maybe the most important question is "Is this even conforming?". Which `::operator new` do we actually use on Windows and does that use our implementation of `{s,g}et_new_handler`? Looking at the code the answer seems to be "not ours", which would make us non-conforming with this patch. Do we have a test to ensure that the new handler is called when trying to allocate too much memory?


**In regards to compatibility:**

- AFAIK the functionality implemented in [`new.cpp`](https://github.com/llvm/llvm-project/blob/main/libcxx/src/new.cpp#L16) is not even used for `VCRuntime`, therefore there shouldn't be any ABI break. and the header change to use `#include <vcruntime_new.h>` is on par with MSVC STL since it also avoids including `<new.h>`, and we avoid using it because that would result in redeclaration warnings (caused by the one function that is exposed by the UCRT `set_new_handler`).

- This patch relies on UCRT functionality which is in line with MSVC STL, and that also means our handler is being called correctly, I've even debugged it using `WinDbg` to make sure it is working correctly while implementing this change.

So in fact, this patch actually makes it compliant by implementing everything as it was supposed to be.

**In regards to testing:** 

Currently the only available [test](https://github.com/llvm/llvm-project/blob/main/libcxx/test/std/language.support/support.dynamic/alloc.errors/set.new.handler/get_new_handler.pass.cpp#L16) is expected to fail (which is not conforming), this patch fixes the functionality and now the test succeeds, as a result I've removed the `XFAIL` in that test. 

As for testing if the handler is being called correctly, I'm not aware of any existing tests for such behaviour.
However, I have tested it locally in different ways by triggering an impossibly large allocation using `new`:
- by setting and retrieving the handler.
- by alternating two different handlers repeatedly in a loop.

But I don't think that this PR should add such tests, they could be added in another PR if needed. 

I hope I have addressed your concerns and recapped some of the reasoning behind this PR, since I ended up implementing more than what was described in the description of the PR in the process of addressing the review comments.

siradam7th commented:
Since https://github.com/llvm/llvm-project/pull/173451 was merged, I've resolved the conflict by moving the changes to the new test file.

philnik777 reviewed (changes requested):
Actually, I've misread it again. We're storing the new handler in two separate places now, so this doesn't work if someone calls `set_new_handler` from MSVC and calls `get_new_handler` from libc++. IoW, the value returned from `get_new_handler` may not be the same as `operator new` would be calling.

This already breaks down if both libc++ and the MSVC STL DLLs are loaded, since there is no guarantee which one is actually used AFAICT. We provide identical symbols in separate DLLs now.

Do we try to load the MSVC STL in other places already possibly? IIRC we have some similar sharing elsewhere. If so, maybe we should just use the MSVC provided functions instead and don't bother providing our own. That should address any concerns I have.

mstorsjo commented:
> Actually, I've misread it again. We're storing the new handler in two separate places now, so this doesn't work if someone calls `set_new_handler` from MSVC and calls `get_new_handler` from libc++. IoW, the value returned from `get_new_handler` may not be the same as `operator new` would be calling.
> 
> This already breaks down if both libc++ and the MSVC STL DLLs are loaded,

In essentially all cases, we have both libc++ and MSVC STL symbols loaded, as libc++ on MSVC still relies on a bunch of ABI symbols provided by the MSVC STL (essentially corresponding to libc++abi in the Itanium world) - either linked dynamically or statically.

> since there is no guarantee which one is actually used AFAICT.

Once linked, there are guarantees (contrary to ELF, each imported symbol is tied to be imported from a specific named DLL). On the linking stage, I guess it depends on which library is provided first to the linker.


> Do we try to load the MSVC STL in other places already possibly? IIRC we have some similar sharing elsewhere. If so, maybe we should just use the MSVC provided functions instead and don't bother providing our own. That should address any concerns I have.

I'm sorry I haven't had time to keep up with this PR thread in detail - but this overall also mirrors my concern. But do I understand things correctly, that we so far have been using the function from vcruntime, and MS STL also reimplements this on top of vcruntime? And we can't use those functions from MS STL, but also need to reimplement it on top of vcruntime separately?

siradam7th committed:
[libc++] add test for 'set_new_handler' when an allocation fails

philnik777 reviewed (commented):
I feel like we should document what our goal for Windows/MSVC is. What sort of compatibility do we want and what should be separated? Do we want to avoid any MSVC headers? In that case we would also want to avoid `<vcruntime_new.h>`. @petrhosek @mstorsjo do you believe separating libc++ completely from MSVC is a goal we should have? Are there people who might not want this separation?

mccakit commented:
> > but MinGW doesn’t support AVX-512
> 
> Can you elaborate on this? Do you mean that GCC for MinGW targets doesn't support AVX-512? I'm not aware of any reason why it wouldn't work with Clang - e.g. through https://github.com/mstorsjo/llvm-mingw/releases.
> 
> (There are of course entirely valid reasons for not wanting to use MinGW - but I don't understand this argument.)

<

You can compile binaries that use avx512 optimizations with mingw triplet, but they will fail at runtime. I don't know the reason behind it, all know are my own runtime errors and PR's on the internet that disable avx512 for windows targets for some libraries.

msvc triple doesn't have this issue, and with the latest wine updates avx512 even works with wine in addition to normal execution.

ldionne commented:
Summarizing the changes needed before we land this:

1. Drop formatting-only changes (e.g. in `global_new_delete.h`)
2. Split `[[nodiscard]]` into a separate PR
3. Add a test for new handler being called on allocation failure
4. Add a release note for LLVM 23
5. Let's document the Windows/MSVC support goals in libcxx/docs/DesignDocs/WindowsSupport.rst (which does not exist today). This should cover the dependency model (using vcruntime as the ABI layer, and our goal to decouple from the MSVC STL). Let's also mention that allowing interoperation between the two (e.g. linking a binary built with libc++ and one built with MSVC STL and expecting they share the same new handler) is not a goal, unless I've misunderstood the discussion on this point (@mstorsjo).

Once those are addressed, I think this is good to go and I'm glad that folks who care about Windows support are aligned that this is the right path forward.

siradam7th committed:
[libc++] add release note for LLVM 23

siradam7th committed:
[libc++] fix missing 'get_new_handler()' for Windows builds

siradam7th committed:
[libc++] remove 'FIXME' and 'XFAIL' for 'get_new_handler()' test

siradam7th committed:
[libc++] implement 'set_new_handler' and 'get_new_handler' for Windows

siradam7th committed:
[libc++] add test for 'set_new_handler' when an allocation fails

frederick-vs-ja commented on libcxx/docs/ReleaseNotes/23.rst:
```diff
@@ -131,5 +131,9 @@ ABI Affecting Changes
 - ``std::exception_ptr`` now uses the same layout in MSVC mode as on other platforms, changing the size from two to one
   pointer.
 
+- ``std::set_new_handler`` and ``std::get_new_handler`` are now implemented by libc++ on Windows using ``VCRuntime``
+  directly instead of implicitly relying on ``msvcprt`` to implement them, therefore the global state of the handler
+  is no longer shared with ``msvcprt`` since libc++ now maintains its own internal state of the handler.
+
```
Oh, I guess we should move this entry to `24.rst`.

siradam7th committed:
[libc++] add release note for LLVM 24

siradam7th commented on libcxx/test/std/language.support/support.dynamic/alloc.errors/set.new.handler/set_new_handler.pass.cpp:
```diff
@@ -6,20 +6,37 @@
+  std::set_new_handler(f3);
+  size_t max_size = std::numeric_limits<size_t>::max();
+  // cap size to (~4GB) on 32-bit, (~1Exabytes) on 64-bit
+  size_t alloc_size = max_size > UINT_MAX ? max_size / 16 : max_size - 1;
```
It doesn't seem to be necessary anymore with SAN disabled, since just assigning `numeric_limits<size_t>::max()` to a variable is enough to stop Clang from issuing a warning.

siradam7th replied:
I turns out GCC minds, but adding `volatile` fixed it.

mccakit commented:
ping @mstorsjo

mstorsjo commented:
> ping @mstorsjo

Not really sure what's needed from me here - it has multiple approvals, including for the libcxx review group itself - so I would expect someone of those who have approved it to merge it.

But I guess I can press the button as well...

mccakit commented:
Finally.. Jesus christ

Zingam commented:
@siradam7th Can you investigate? One of the failures is related to this PR IMO.

siradam7th commented:
> @siradam7th Can you investigate? One of the failures is related to this PR IMO.

@Zingam This CI seems to be different than what is running on PRs since this was not happening before.
The test file `set_new_handler.pass.cpp‎` must disable sanitizers else they would prevent it from running successfully, looking at the logs at https://lab.llvm.org/buildbot/#/builders/55/builds/32041 it looks like that CI enables an extra SAN `hwasan` that must be disabled too.

mstorsjo commented:
> > @siradam7th Can you investigate? One of the failures is related to this PR IMO.
> 
> @Zingam This CI seems to be different than what is running on PRs since this was not happening before. The test file `set_new_handler.pass.cpp‎` must disable sanitizers else they would prevent it from running successfully, looking at the logs at https://lab.llvm.org/buildbot/#/builders/55/builds/32041 it looks like that CI enables an extra SAN `hwasan` that must be disabled too.

Can you put up a PR that adds `hwasan` (which does seem to be mentioned in a few other existing tests) to the `UNSUPPORTED:` list in that test?

siradam7th commented:
@mstorsjo I just did https://github.com/llvm/llvm-project/pull/217086/
You have to run the same CI that is failing since it doesn't seem to have the same configuration as the one that runs on PRs before you can merge it.

Zingam commented:
> > @siradam7th Can you investigate? One of the failures is related to this PR IMO.
> 
> @Zingam This CI seems to be different than what is running on PRs since this was not happening before. The test file `set_new_handler.pass.cpp‎` must disable sanitizers else they would prevent it from running successfully, looking at the logs at https://lab.llvm.org/buildbot/#/builders/55/builds/32041 it looks like that CI enables an extra SAN `hwasan` that must be disabled too.

I don't know how to explain this correctly, let's say there are different groups of build bots out there that we need to watch for. 

libc++ pre-merge ones https://github.com/llvm/llvm-project/pull/150182/checks

LLVM post-merge others https://lab.llvm.org/buildbot/#/builders

Chromium ones https://ci.chromium.org/p/fuchsia/g/llvm/console

Zingam commented:
> > > @siradam7th Can you investigate? One of the failures is related to this PR IMO.
> > 
> > 
> > @Zingam This CI seems to be different than what is running on PRs since this was not happening before. The test file `set_new_handler.pass.cpp‎` must disable sanitizers else they would prevent it from running successfully, looking at the logs at https://lab.llvm.org/buildbot/#/builders/55/builds/32041 it looks like that CI enables an extra SAN `hwasan` that must be disabled too.
> 
> I don't know how to explain this correctly, let's say there are different groups of build bots out there that we need to watch for.
> 
> libc++ pre-merge ones https://github.com/llvm/llvm-project/pull/150182/checks
> 
> LLVM post-merge others https://lab.llvm.org/buildbot/#/builders
> 
> Chromium ones https://ci.chromium.org/p/fuchsia/g/llvm/console

Failing one of these may result in your commit being reverted that's why I pinged you.

mstorsjo commented:
> You have to run the same CI that is failing since it doesn't seem to have the same configuration as the one that runs on PRs before you can merge it.

Even if the bot owner hasn't test run it, I think the fix seems safe and reasonable enough that I went ahead and landed it.
